src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextUI.java

```diff
@@ -1908,6 +1908,12 @@ public final void propertyChange(PropertyChangeEvent evt) {
             } else if ("componentOrientation" == propertyName) {
                 // Changes in ComponentOrientation require the views to be
                 // rebuilt.
+                Document document = editor.getDocument();
+                final String I18NProperty = "i18n";
+                if (ComponentOrientation.RIGHT_TO_LEFT == newValue
+                    && ! Boolean.TRUE.equals(document.getProperty(I18NProperty))) {
```

Comment:
Please add a comment similar to the text in the AbstractDocument.handleInsertString, otherwise, it is unclear why we need to enable "complex layout".